### ISSUE TITLE
Performance issues since Reaper 6.71 - slow / laggy Automation recording

### ISSUE THREAD

**Issue by Fleeesch** [helgoboss/realearn#787]
Greetings,

I'm getting some performance issues with ReaLearn and Reaper starting with version 6.71 (Windows 10). Just wondering if I'm the only one getting them.

Everything is smooth as Butter with Reaper 6.69 and ReaLearn 2.14.0, even on my lower-spec devices. I make the jump to 6.71 and automation recording with ReaLearn becomes extremly laggy.

**[Link to Reaper Foum thread](https://forum.cockos.com/showthread.php?t=273694)**

### Issues
- extremly slow/delayed automation recording
- unusually high CPU usage during automation recording
- Reaper GUI is lagging heavily during automation recording
- constant top menu bar flashing / redrawing
- large amounts of undo points created during recording

### What I've tried
- fresh install of Reaper
- turning off every feature possible that is not required
- using Reapers per-plugin compatibilty settings
- typical system checks (faulty drivers, background software, etc.)
- **reverting to older ReaLearn versions _-> gets better with 2.14.0-pre.9, but still janky_**

### Example Project

[automation_error_6-71_example.zip](https://github.com/helgoboss/realearn/files/10321973/automation_error_6-71_example.zip)
Just press play, ReaLearn then translates individual 14-Bit CC sweeps into volume automation data.

### Visuals

6.69 behvaiour (normal)
![bugreport_221213_automation_error_6 69](https://user-images.githubusercontent.com/96627277/210009600-f9539e61-12bb-4c77-85c5-65db338cadba.gif)

6.71 behaviour (bad)
![bugreport_221213_automation_error_6 71](https://user-images.githubusercontent.com/96627277/210009607-74cb9672-d1ce-4ace-819e-c250637eec0f.gif)

```
* * Specs * *

Reaper
- Official Realeases: 6.69, 6.71, 6.73
- 64-Bit, Windows

Desktop (custom)
- Win: 10 64-Bit (10.0.19045 Build 19045)
- i5: 2400 CPU @ 3.10GHz
- RAM: 8GB
- GPU: Nvidia Geforce GT 1030

Laptop (Lenovo T480)

- OS: Win 10 64-Bit (10.0.19045 Build 19045)
- CPU: i7-8550U
- RAM: 16GB
- GPU: Intel HD Graphics 620 / NVIDIA GeForce MX150
```

**Comment by PucksEvilTwin** [comment]
Your example project behaves the same on my system.

- Win 10
- i7 4790k
- Ram: 16GB
- GPU: Nvidia Geforce 1060

However, recording automation how I normally do it, using OSC via TouchOSC on a Touchscreen monitor, everything is still as smooth as butter.

**Comment by Fleeesch** [comment]
Are you using ReaLearn to map the OSC messages or Reapers internal OSC control support (the one under the "Control/OSC/web" tag in preferences)?

I'm only getting the slowdown issue with ReaLearn, recording automation with Reapers internal Mackie Control Universal protocol works without a problem.

**Comment by PucksEvilTwin** [comment]
Im using Realearn’s OSC implementation.  Sorry, I figured that would be obvious but it makes sense to be ultra clear.  In interest of being clear, I’m using touch automation here when I record my automation on my setup.

**Comment by helgoboss** [comment]
Interesting. @Fleeesch Did you try older ReaLearn versions with the latest REAPER? Same effect? What about REAPER 6.70? Maybe the REAPER changelog can reveal a suspicious change: https://www.reaper.fm/whatsnew.txt

Sorry, I'm a bit busy at the moment, not enough time to look into it.

**Comment by Fleeesch** [comment]
@helgoboss Here's a short documentation:
- Reaper 6.69 & ReaLearn v2.3+
  - _All good_
- Reaper 6.70 & ReaLearn v2.3+
  - _Reduced automation data, automation writing introduces UI lag_
- Reaper 6.71+ & ReaLearn v2.3+
  - _Reduced automation data, automation writing introduces UI lag_
- Reaper 6.71+ & Realearn v2.14.0-pre.10+
  - _Heavy slowdown of UI, top menu bar flashing, tons of undo points created_


[There's another user having UI issues with a recent version of Reaper](https://forum.cockos.com/showthread.php?t=274327) , unrelated to ReaLearn. **The same kind of GUI slowdown also occurs when you drag a fader with the mouse to write automation data in realtime, on a fresh install.** Sounds like an issue with Reaper itself. OP also mentoined a huge imrovment in GUI reliability with Reaper 6.69.


I went through the changelogs of the official Reaper releases, the only entry that seems to be GUI-related is this one:
```
Reaper Release v6.70rc1 - November 14 2022
+ Envelopes: improve redraw of envelope panel knob with touch automation
```

**Comment by helgoboss** [comment]
Thanks for the detailed testing, Fleeesch :)

> Reaper 6.71+ & Realearn v2.14.0-pre.10+
> * _Heavy slowdown of UI, top menu bar flashing, tons of undo points created during automation writing_

Oh, so at least this point seems to be associated with changes in ReaLearn. Any chance you can pinpoint which release? 2.3+ to 2.14 is a large gap. Or did you mean 2.13+ instead of 2.3+?

**Comment by Fleeesch** [comment]
Glad to help. That's a typo indeed, I've meant to write 2.13.0.

From Reaper 6.71 on the heavy issues (top menu flashing) are introduced with **Realearn v2.14.0-pre.10**. The top menu flashing disappears if you turn off the "Show last undo point in menu bar", but the faulty automation data remains. Prior to Reaper 6.71 I couldn't find any difference in behaviour between ReaLearn versions starting from v2.13.0.

All the weird stuff that's happening seems to be closely related to Reapers GUI drawing. Things get a lot better when I scale down the Reaper window a lot. I initially thought the bottleneck might be my UHD display and some bad GPU compatibility issue, but the behaviour is almost identical on my laptop (Full HD, windows 125% scale).


**Here's a demonstration of the menu flashing / undo point creation on a fresh Reaper install:**

![bugreport_221213_automation_error_flashing_6 71](https://user-images.githubusercontent.com/96627277/210903378-60dfdff4-da2e-4481-a381-c0df315d4568.gif)

**Comment by RattlesnakeLodge** [comment]
To chime in, I've noticed a similar issue, using 2.14 on Reaper 6.73. I noticed the flashy/flickering menu bar when doing simple like adjusting track volume via fader, or track pan via v-pot. No automation, just basic track adjustments.

I looked into undo history, and it's populated with a bunch of tiny increments of track volume adjustments, track pan adjustments, etc. It's quite a lot of undo points. Looking at the performance meter while making fader movement, CPU and disk usage do jump up a bit, likely from the writing of all those undo points.

Flipping into CSI, the issue is not present. Track volume adjustments have only one undo point.

Also, flipping back into ReaLearn, changing the target from Track Volume to a Reaper action (adjust track volume via midi cc/osc) only creates one undo point per adjustment.
